### PR TITLE
p2p: prevent connections from same ip

### PR DESCRIPTION
**Pull request by ebuchman**
#1264 

TODO: fix and write tests

* [ ] Updated all relevant documentation in docs
* [ ] Updated all code comments where relevant
* [ ] Wrote tests
* [ ] Updated CHANGELOG.md

### CODE REVIEW

File: p2p/peer_set.go [hunk 1]
```diff
@@ -76,6 +92,7 @@ func (ps *PeerSet) Get(peerKey ID) Peer {
 func (ps *PeerSet) Remove(peer Peer) {
 	ps.mtx.Lock()
 	defer ps.mtx.Unlock()
+	delete(ps.lookupIP[peer.RemoteIP()])
```
Comment (odeke-em) [inline p2p/peer_set.go]: Does this compile?

File: p2p/peer_set.go [hunk 2]
```diff
@@ -61,6 +68,15 @@ func (ps *PeerSet) Has(peerKey ID) bool {
 	return ok
 }
 
+// HasIP returns true iff the PeerSet contains
+// the peer referred to by this IP address.
+func (ps *PeerSet) HasIP(peerIP string) bool {
```
Comment (melekes) [inline p2p/peer_set.go]: maybe we should use `net.IP` here?

Comment (xla) [reply p2p/peer_set.go]: :+1:

File: p2p/peer.go [hunk 3]
```diff
@@ -45,6 +46,15 @@ func (pc peerConn) ID() ID {
 	return PubKeyToID(pc.conn.(*tmconn.SecretConnection).RemotePubKey())
 }
 
+// Return the IP from the connection RemoteAddr
+func (pc peerConn) RemoteIP() string {
+	host, _, err := net.SplitHostPort(pc.conn.RemoteAddr().String())
+	if err != nil {
+		panic(err)
```
Comment (xla) [inline p2p/peer.go]: Is this case worth a panic?

File: p2p/peer.go [hunk 4]
```diff
@@ -45,6 +46,15 @@ func (pc peerConn) ID() ID {
 	return PubKeyToID(pc.conn.(*tmconn.SecretConnection).RemotePubKey())
 }
 
+// Return the IP from the connection RemoteAddr
+func (pc peerConn) RemoteIP() string {
```
Comment (xla) [inline p2p/peer.go]: If we truly want to return an ip address I think we should do a lookup in case `RemoteAddr()` is a dns name.

**Comment by ebuchman** [comment]
:o same test failure as anton is having in https://github.com/tendermint/tendermint/pull/1574#issuecomment-390768376

**Comment by ebuchman** [comment]
Oh well, here at least presumably it's becuase they all have the same IP, so they disconnect when they realize thaT?

**Comment by ebuchman** [comment]
maybe what we need is some kind of "filter" object that we can set that determines what peers we disconnect to ... that way in tests we don't need to apply the same level of filters so we can not worry about connecting to self when we're trying to test something else ... ?

File: p2p/switch.go [hunk 5]
```diff
@@ -534,6 +534,8 @@ func (sw *Switch) addPeer(pc peerConn) error {
 		return err
 	}
 
+	// dont connect to multiple peers on the same IP
```
Comment (ebuchman) [inline p2p/switch.go]: out of place ?

File: p2p/peer.go [hunk 6]
```diff
@@ -45,6 +50,35 @@ func (pc peerConn) ID() ID {
 	return PubKeyToID(pc.conn.(*tmconn.SecretConnection).RemotePubKey())
 }
 
+// Return the IP from the connection RemoteAddr
+func (pc peerConn) RemoteIP() net.IP {
+	if len(pc.ips) > 0 {
+		return pc.ips[0]
+	}
+
+	if pc.conn == nil || pc.conn.RemoteAddr().String() == "pipe" {
```
Comment (ebuchman) [inline p2p/peer.go]: is this only nil during test? we should add comment either way that this little section is just for tests.

File: p2p/peer.go [hunk 7]
```diff
@@ -45,6 +50,35 @@ func (pc peerConn) ID() ID {
...
+		pc.ips = []net.IP{
+			net.IP{172, 16, 0, byte(atomic.AddUint32(&testIPSuffix, 1))},
+		}
+
+		return pc.ips[0]
+	}
+
+	host, _, err := net.SplitHostPort(pc.conn.RemoteAddr().String())
```
Comment (ebuchman) [inline p2p/peer.go]: Does RemoteAddr return host names or only IPs ?

File: p2p/peer.go [hunk 8]
```diff
@@ -37,6 +41,7 @@ type peerConn struct {
 	persistent bool
 	config     *PeerConfig
 	conn       net.Conn // source connection
+	ips        []net.IP
```
Comment (ebuchman) [inline p2p/peer.go]: are you thinking we'll use this for something ? looks like we only ever use the [0]

File: p2p/switch.go [hunk 9]
```diff
@@ -564,20 +566,21 @@ func (sw *Switch) addPeer(pc peerConn) error {
...
 	}
 
 	// Avoid duplicate
 	if sw.peers.Has(peerID) {
-		return ErrSwitchDuplicatePeer
+		return ErrSwitchDuplicatePeerID{peerID}
+	}
+
+	// check ips for both the connection addr and the self reported addr
+	if sw.peers.HasIP(pc.RemoteIP()) {
```
Comment (ebuchman) [inline p2p/switch.go]: this is the connection address. Maybe we should check the self-reported address from the NodeInfo as well ?

File: p2p/peer_set_test.go [hunk 10]
```diff
@@ -112,25 +124,38 @@ func TestPeerSetAddDuplicate(t *testing.T) {
...
-	wantErrCount, gotErrCount := n-1, errsTally[ErrSwitchDuplicatePeer]
+	wantErrCount, gotErrCount := n-1, errsTally["duplicateID"]
 	assert.Equal(t, wantErrCount, gotErrCount, "invalid ErrSwitchDuplicatePeer count")
 
-	wantNilErrCount, gotNilErrCount := 1, errsTally[nil]
+	wantNilErrCount, gotNilErrCount := 1, errsTally["other"]
 	assert.Equal(t, wantNilErrCount, gotNilErrCount, "invalid nil errCount")
 }
 
+func TestPeerSetAddDuplicateIP(t *testing.T) {
```
Comment (ebuchman) [inline p2p/peer_set_test.go]: TODO

**Comment by ebuchman** [comment]
Looks good. Some minor comments, but mostly just need to address the test issue before merge.

Note the test failure here actually has a switch with no peers and makes sense why its not working ( I think ). The fail in #1574 might be different

**Comment by xla** [comment]
@ebuchman Thanks for the thorough review.

**Comment by xla** [comment]
@ebuchman @melekes green at last and ready for a final review.

**Comment by ebuchman** [comment]
LGTM!

**Pull request merged**